### PR TITLE
[Swap][Tests][KEP-2400] Add swap serial stress tests, improve NodeConformance tests and adapt NoSwap behavior

### LINKED ISSUE #120798: [KEP-2400] [Swap]: SwapConformance e2e testing
### What would you like to be added?

This issue is a follow-up to https://github.com/kubernetes/kubernetes/issues/119430 and https://github.com/kubernetes/kubernetes/issues/105025.

Currently, we have swap e2e tests. However, these tests only check that the cgroup knobs (e.g. `memory.swap.max`) had been configured properly.

This issue focuses on e2e tests that would actually get swap to kick in and would prove that over-committing the node is possible.

### Why is this needed?

KEP https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2400-node-swap

### PR DESCRIPTION
**Pull request by iholder101**
#### What type of PR is this?
/kind feature

Fixes https://github.com/kubernetes/kubernetes/issues/120798

#### What this PR does / why we need it:
This PR does the following:
1. Havily refactors and cleans swap node-e2e tests.
2. Adapt swap NodeConformance test to https://github.com/kubernetes/kubernetes/pull/122745 which drops `UnlimitedSwap` and introduces `NoSwap` behavior.
3. Adds serial tests for swap.

To expand on (3), two serial tests are now added which validate the following scenarios:
1. `"should be able over-commit the node memory"` - a pod is deployed which stresses memory. The test expects that once the pod uses all of the node's memory some of its memory would be swapped away.
2. `"should be able to use more memory than memory limits"` - a pod with memory limits stresses memory. Eventually it goes beyond its memory limits and instead of being OOM killed some of its memory is swapped away keeping it alive.

#### Special notes for your reviewer:
The original version of this PR was here: https://github.com/kubernetes/kubernetes/pull/120430

Due to unfortunate situations I had to be out of work for a while, therefore my PR was closed and continued here: https://github.com/kubernetes/kubernetes/pull/122175

Fortunately I was able to come back to work and therefore the PR is now continued here.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md
```

### CODE REVIEW

**Comment by iholder101** [comment]
PR is not ready to review yet.

/uncc @dchen1107 @andrewsykim

**Comment by kannon92** [comment]
/cc

when it is ready

**Comment by kannon92** [comment]
We have crio ci failures in upstream.

https://testgrid.k8s.io/sig-node-cri-o#ci-crio-cgroupv1-node-e2e-features
https://testgrid.k8s.io/sig-node-cri-o#ci-crio-cgroupv1-node-e2e-conformance
https://testgrid.k8s.io/sig-node-cri-o#ci-crio-cgroupv2-node-e2e-conformance

I think you are impacted by this.

File: test/e2e_node/swap_test.go [hunk 1]
```diff
@@ -38,49 +44,215 @@ import (
...
-		expectedSwapLimit := calcSwapForBurstablePod(f, pod)
-		expectLimitedSwap(f, pod, expectedSwapLimit)
-	},
-		ginkgo.Entry("QOS Best-effort", v1.PodQOSBestEffort, false),
-		ginkgo.Entry("QOS Burstable", v1.PodQOSBurstable, false),
-		ginkgo.Entry("QOS Burstable with memory request equals to limit", v1.PodQOSBurstable, true),
-		ginkgo.Entry("QOS Guaranteed", v1.PodQOSGuaranteed, false),
-	)
+			switch swapBehavior := getSwapBehavior(); swapBehavior {
+			case types.LimitedSwap:
```
Comment (kannon92) [inline test/e2e_node/swap_test.go:75]: I think there is no way this code will be hit in conformance test suite. We have to test the default path because we can't change the kubelet config in non-serial tests.

Comment (iholder101) [reply test/e2e_node/swap_test.go:75]: > I think there is no way this code will be hit in conformance test suite.

Yes, this is true.
My reasoning behind this is that we might change defaults / what node conformance is using one day, then it would be less work. IOW, the test doesn't care which swap behavior you choose to run, it just makes sure that the right behavior occurs.

If you think it's better, we can instead expect that the behavior is indeed `NoSwap` and fail otherwise.

> we can't change the kubelet config in non-serial tests.

Right, and we don't :)

Comment (kannon92) [reply test/e2e_node/swap_test.go:75]: > If you think it's better, we can instead expect that the behavior is indeed NoSwap and fail otherwise.

I would go this route. Less code is better in this scenario. I really don't see us changing the defaults and if we do I would rather have the tests fail and people update them.

Comment (iholder101) [reply test/e2e_node/swap_test.go]: Done :+1:

**Comment by iholder101** [comment]
@kannon92

NodeConformance tests have passed on `e2e-containerd`.
The serial lanes failed for seemingly unrelated reasons, but swap tests have passed there as well [[1](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/123557/pull-kubernetes-node-swap-fedora-serial/1767575322049384448)][[2](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/123557/pull-kubernetes-node-swap-ubuntu-serial/1767575321990664192)].

The PR is ready to review :) PTAL

**Comment by kannon92** [comment]
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial

Looks like these tests are timing out. Maybe your test is making them exceed the time limit allocated to them.

**Comment by iholder101** [comment]
> /test pull-kubernetes-node-swap-fedora-serial
> /test pull-kubernetes-node-swap-ubuntu-serial

(re-launching tests as I've pushed a minor change to satisfy the linter)
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial

> Looks like these tests are timing out. Maybe your test is making them exceed the time limit allocated to them.

Cloud be.
I'll keep an eye on that.

**Comment by iholder101** [comment]
Unrelated failure [[1](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/123557/pull-kubernetes-node-swap-fedora-serial/1767611214742949888)].

/test pull-kubernetes-node-swap-fedora-serial

**Comment by kannon92** [comment]
Looking at the serial logs, it looks like the agnhost container is being OOMed for fedora: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/123557/pull-kubernetes-node-swap-fedora-serial/1767625656390127616/artifacts/n1-standard-2-fedora-coreos-39-20240210-3-0-gcp-x86-64-436ed8ef/serial-1.log

I have been trying to chase down a really nasty bug with serial jobs that you seem to be hitting with ubuntu. 

https://github.com/kubernetes/kubernetes/issues/123589

Looking at those serial logs, I am seeing that we are hitting a OOM also for dd.

**Comment by iholder101** [comment]
> Looking at the serial logs, it looks like the agnhost container is being OOMed for fedora: [storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/123557/pull-kubernetes-node-swap-fedora-serial/1767625656390127616/artifacts/n1-standard-2-fedora-coreos-39-20240210-3-0-gcp-x86-64-436ed8ef/serial-1.log](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/123557/pull-kubernetes-node-swap-fedora-serial/1767625656390127616/artifacts/n1-standard-2-fedora-coreos-39-20240210-3-0-gcp-x86-64-436ed8ef/serial-1.log)
> 
> I have been trying to chase down a really nasty bug with serial jobs that you seem to be hitting with ubuntu.
> 
> #123589
> 
> Looking at those serial logs, I am seeing that we are hitting a OOM also for dd.

So I've been tweaking things a bit, and all of the tests passed on a GCE instance (identical to the one we use on CI) for multiple times. Let's see if CI is on our side.

/test pull-kubernetes-node-swap-fedora
/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial

**Comment by iholder101** [comment]
Swap tests passed on both lanes, although they have failed. [[1](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/123557/pull-kubernetes-node-swap-fedora-serial/1767858931914248192)][[2](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/123557/pull-kubernetes-node-swap-ubuntu-serial/1767858931855527936)].

@kannon92 do you mind help me figure out the flakiness of these lanes before we add the tests presented in this PR? Since the tests passed for 20 times in a row on an identical GCE instance, I wonder whether the lane is flaky by nature or if the newly introduced tests actually hurt its stability.

/test pull-kubernetes-node-swap-fedora-serial
/test pull-kubernetes-node-swap-ubuntu-serial

File: test/e2e/nodefeature/nodefeature.go [hunk 2]
```diff
@@ -100,6 +100,10 @@ var (
 	// TODO: document the feature (owning SIG, when to use this feature for a test)
 	RuntimeHandler = framework.WithNodeFeature(framework.ValidNodeFeatures.Add("RuntimeHandler"))
 
+	// Added to test Swap Feature
+	// This label should be used when testing KEP-2400 (Node Swap Support)
+	Swap = framework.WithNodeFeature(framework.ValidNodeFeatures.Add("NodeSwap"))
```
Comment (kannon92) [inline test/e2e/nodefeature/nodefeature.go:105]:
```suggestion
	Swap = framework.WithNodeFeature(framework.ValidNodeFeatures.Add("SwapConformance"))
```

I think.

Comment (iholder101) [reply test/e2e/nodefeature/nodefeature.go:105]: Not exactly.
The label you're pointing to here is for the NodeSwap feature gate. Instead I've added a `WithSwapConformance()` function to add the swap conformance label. Look at the [latest commit](https://github.com/kubernetes/kubernetes/pull/123557/commits/8a9a9617293551bddee0d4268fd7d0a71cedc1f5) for the implementation.

**Comment by iholder101** [comment]
@kannon92 Looking good! :tada: 

/test pull-kubernetes-node-swap-fedora-serial

**Comment by iholder101** [comment]
The fact that `swap-fedora-serial` is failing even without running swap conformance tests shows it is probably flaky and that this PR is not the cause.

Trying one last time.

/test pull-kubernetes-node-swap-fedora-serial

File: test/e2e/nodefeature/nodefeature.go (hunk 2 above)
Comment (kannon92) [reply test/e2e/nodefeature/nodefeature.go:105]: Thats not true. This is just the label that we use for FOCUS.

Comment (iholder101) [reply test/e2e/nodefeature/nodefeature.go:105]: Right. That's what I meant. Let me rephrase.

We use `"NodeSwap"` at the upper `SIGDescribe`, meaning that this label appears for every swap test.
The new `"SwapConformance"` label introduced in this PR is applied only for the swap conformance tests.

This way:
- A job that focuses on NodeSwap runs all swap tests.
- A job that focuses on Serial will NOT run swap conformance.
- A job that focuses on SwapConformance will run swap conformance tests only.

Does this make sense? For more info look at the latest commit.

Comment (kannon92) [reply test/e2e/nodefeature/nodefeature.go:105]: So does "NodeSwap runs all swap tests" means all swap tests without SwapConformance?

File: test/e2e_node/swap_test.go [hunk 3]
```diff
@@ -38,49 +45,229 @@ import (
...
+						if err != nil {
+							return err
+						}
+
+						if swapUsage.IsZero() {
+							return fmt.Errorf("swap usage is zero")
+						}
+
+						// Better to delete the stress pod ASAP to avoid node failures
+						_ = podClient.Delete(context.Background(), stressPod.Name, metav1.DeleteOptions{})
```
[the remark this answers is not in the document]
Comment (iholder101) [reply test/e2e_node/swap_test.go]: > I don't think we should call this more than once?
> Could we move it out of the eventually?

It is being called only once before the Eventually clause returns no err (nil).
I basically wanted to delete the pod ASAP to avoid compromising the node, but there's no reason to do so right after the eventually clause ends.

Thanks @kannon92! I agree, the code is much clearer this way.
Done.

> And can we assert there is no error?

Sure. Done.

File: test/e2e/nodefeature/nodefeature.go (hunk 2 above)
Comment (iholder101) [reply test/e2e/nodefeature/nodefeature.go:105]: > So does "NodeSwap runs all swap tests" means all swap tests without SwapConformance?

No. Every swap test has the NodeSwap label. It makes sense, as if a job focuses only on NodeSwap it would be expected to run all swap related tests.

* All swap test has a NodeSwap label.
  * All swap node-conformance tests **also** have the NodeConformance label.
  * All swap swap-conformance tests **also** have the SwapConformance label.

Makes sense?

Please look at the code here which describes this:
https://github.com/iholder101/kubernetes/blob/635f936fcc3ffe902ef228defef4cfec367b2fbc/test/e2e_node/swap_test.go#L58-L85

Comment (iholder101) [reply test/e2e/nodefeature/nodefeature.go:105]: In short:
NodeConformance job would run only non-swap-conformance tests.
SwapConformance job would run only swap-conformance tests.

Comment (kannon92) [reply test/e2e/nodefeature/nodefeature.go:105]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-testing/3041-node-conformance-and-features could be a good read for you. 

NodeConformance is meant to be a test that can be run in all environments and not tied to any feature. SwapConformance seems to be meaningless and it is really a feature test of swap.

I'm going to tag @pohly and see what we think about `SwapConformance`.

Comment (iholder101) [reply test/e2e/nodefeature/nodefeature.go:105]: I just think that a test focuses on NodeSwap should run both node-conformance and non-node-conformance tests if they are related to the NodeSwap feature.

WDYT about replacing `SwapConformance` with something like `SerialSwap`?

In any case, if there's a strong push-back, I can add the NodeSwap label to serial tests only.

Comment (kannon92) [reply test/e2e/nodefeature/nodefeature.go:105]: I think having a swap enabled node run the node-conformance tests covers that case. So you could remove all swap related tests from the serial and non-serial swap jobs.

And in your "SwapConformance" jobs, you can just run all the tests that involve the swap feature. 

And then all the tests of the feature can be run on the serial node. You can test both NoSwap and LimitedSwap on serial because you can change the config.

I don't see a reason to add "SerialSwap". Serial is mostly just a implementation detail for the feature tests because we need to change from the default.

Comment (iholder101) [reply test/e2e/nodefeature/nodefeature.go:105]: @kannon92 
Alright, so let's summarize your suggestion to make sure we're on the same page here.

What you suggest is to remove `SwapConformance` from here:
https://github.com/iholder101/kubernetes/blob/635f936fcc3ffe902ef228defef4cfec367b2fbc/test/e2e_node/swap_test.go#L85

Then the swap-conformance jobs would focus on `NodeSwap` (i.e. `nodefeature.Swap`)?

This way node-conformance would run the non-stress swap tests
And swap-conformance would run ALL swap-related tests?

I'm fine with this approach

Comment (kannon92) [reply test/e2e/nodefeature/nodefeature.go:105]: Ideally, it would be nice to drop SwapConformance label entirely and just have all swap tests under NodeSwap. I think you titled the test-infra job for SwapConformance to target SwapConformance so we may need to change that first.

Comment (iholder101) [reply test/e2e/nodefeature/nodefeature.go:105]: @kannon92 Sure. Here's the PR: https://github.com/kubernetes/test-infra/pull/32276.

What are your thoughts on https://github.com/kubernetes/kubernetes/pull/123557#discussion_r1530313746?

Comment (kannon92) [reply test/e2e/nodefeature/nodefeature.go:105]: > BTW, I'm not sure what value we gain from running pull-kubernetes-node-swap-fedora-serial and pull-kubernetes-node-swap-ubuntu-serial, since they use NoSwap.

I think the goal was to make sure that enabling swap wouldn't break the conformance tests. I guess that since we defaulted from UnlimitedSwap to NoSwap we are only testing if swap will impact any of the tests.
> Isn't it (almost) similar to running the tests without swap enabled?
Don't we want to use LimitedSwap for these jobs?

We could but its not the default and changing a default in a non serial test is not really allowed. Changing a config requires a restart of kubelet..

We could change the serial jobs to use LimitedSwap but not sure what value that would have.

I could see us dropping those lanes eventually but I say we should keep them in case.

Comment (iholder101) [reply test/e2e/nodefeature/nodefeature.go:105]: > > BTW, I'm not sure what value we gain from running pull-kubernetes-node-swap-fedora-serial and pull-kubernetes-node-swap-ubuntu-serial, since they use NoSwap.
> 
> I think the goal was to make sure that enabling swap wouldn't break the conformance tests. I guess that since we defaulted from UnlimitedSwap to NoSwap we are only testing if swap will impact any of the tests.
> 
> > Isn't it (almost) similar to running the tests without swap enabled?
> > Don't we want to use LimitedSwap for these jobs?
> 
> We could but its not the default and changing a default in a non serial test is not really allowed. Changing a config requires a restart of kubelet..

Right. The lanes I've mentioned are serial.

> We could change the serial jobs to use LimitedSwap but not sure what value that would have.
> 
> I could see us dropping those lanes eventually but I say we should keep them in case.

I just think that at this point we have certain confidence in the fact that swap enabled on the node doesn't break serial tests. Running these tests with LimitedSwap would give raise our confidence to a much higher level since we'd know that not only a node with swap enabled doesn't break stuff, but actually using swap with LimitedSwap keeps the node stable.

Just a thought though, not a must.

**Comment by iholder101** [comment]
@kannon92 could you please have another look now that https://github.com/kubernetes/test-infra/pull/32276 is merged and the swap conformance label is removed?

**Comment by kannon92** [comment]
tests look good now

/lgtm

**Comment by kannon92** [comment]
/triage accepted
/priority important-soon

Let's get this into 1.31 as soon as it opens.

**Comment by iholder101** [comment]
@mrunalp Would you mind taking a look?

**Comment by iholder101** [comment]
> /retest

Hey @SergeyKanzhelev!
Please note that the regular `pull-kubernetes-node-swap-*-serial` lanes are flaky.

We've introduced the new `pull-kubernetes-node-swap-conformance-*-serial` lanes, which are stable and run only swap-stress related tests, for that purpose. These tests pass continuously on this PR.

**Comment by pohly** [comment]
Is someone working on fixing this flakiness?

Fixing flakes should have more priority than adding more tests or features.

**Comment by iholder101** [comment]
> Is someone working on fixing this flakiness?
> 
> Fixing flakes should have more priority than adding more tests or features.

While I agree it's important, I think it's out-of-scope for this PR.

This PR does not affect the tests that these lanes are running, but rather adds new serial stress tests to run under `pull-kubernetes-node-swap-conformance-*-serial` lanes.

**Comment by kannon92** [comment]
@pohly We are working on fixing flakyness in #sig-node-ci but bandwidth is limited. I don't think we can block all features in sig-node until we have a green CI. 

Fixing flakes is complicated because each test can have a different reason. For example, we have a flake in our eviction tests (for at least two years) and we spent at least a few months working on fixing it. It turns out that there was a limitation in our cadvisor stats code which was reporiting wrong processes. We debated about fixing at the cadvisor level or at the stats level.

File: test/e2e_node/swap_test.go [hunk 4]
```diff
@@ -38,49 +45,231 @@ import (
...
+	cgroupV2MemoryCurrentUsageFile = "memory.current"
 )
 
-var _ = SIGDescribe("Swap", framework.WithNodeConformance(), "[LinuxOnly]", func() {
-	f := framework.NewDefaultFramework("swap-test")
+var (
+	noLimits *resource.Quantity = nil
+)
+
+var _ = SIGDescribe("Swap", "[LinuxOnly]", nodefeature.Swap, func() {
```
[the remark this answers is not in the document]
Comment (haircommander) [reply test/e2e_node/swap_test.go:58]: nevermind, that's what the variable is

**Comment by haircommander** [comment]
/lgtm

serial tests are flaking but the new test passed, thanks @iholder101

**Comment by rphillips** [comment]
lgtm as well

File: test/e2e_node/swap_test.go [hunk 5]
```diff
@@ -197,57 +406,137 @@ func expectLimitedSwap(f *framework.Framework, pod *v1.Pod, expectedSwapLimit in
...
+	}
+
+	ginkgo.By("memory usage found: " + outputStr + " bytes")
+
+	return parseBytesStrToQuantity(outputStr)
+}
+
+// Sets memory request or limit can be null, then it's dismissed.
+// Sets the same value for all containers.
+func setPodMmoryResources(pod *v1.Pod, memoryRequest, memoryLimit *resource.Quantity) {
```
Comment (SergeyKanzhelev) [inline test/e2e_node/swap_test.go:522]: is misspelling of the name intentional?

File: test/e2e_node/swap_test.go [hunk 6]
```diff
@@ -38,49 +45,231 @@ import (
...
+				}
+
+				getStressPod := func(stressSize *resource.Quantity) *v1.Pod {
+					pod := getStressPod(f, stressSize, stressMemAllocSize)
+					pod.Spec.NodeName = nodeName
+
+					return pod
+				}
+
+				ginkgo.It("should be able over-commit the node memory", func() {
```
Comment (SergeyKanzhelev) [inline test/e2e_node/swap_test.go:182]: minor: over-commit sounds like we can request more than node allows. I'd call it should be able to use more than the node memory capacity

File: test/e2e_node/swap_test.go [hunk 7]
```diff
@@ -38,49 +45,231 @@ import (
...
+				expectLimitedSwap(f, pod, expectedSwapLimit)
+			},
+				ginkgo.Entry("QOS Best-effort", v1.PodQOSBestEffort, false),
+				ginkgo.Entry("QOS Burstable", v1.PodQOSBurstable, false),
+				ginkgo.Entry("QOS Burstable with memory request equals to limit", v1.PodQOSBurstable, true),
+				ginkgo.Entry("QOS Guaranteed", v1.PodQOSGuaranteed, false),
+			)
+		})
+
+		// These tests assume the following, and will fail otherwise:
```
Comment (SergeyKanzhelev) [inline test/e2e_node/swap_test.go:119]: thank you for those. What swappiness value did you use in your configs? is it something we need to take into consideration. I understand that kernel may not be super efficient in swapping some times. So depending on how fast we write memory, test may start flaking.

The soft memory limit in cgroupv2 is something that I think may improve swap reliability.

File: test/e2e_node/swap_test.go (hunk 5 above)
Comment (iholder101) [reply test/e2e_node/swap_test.go]: It isn't, thanks for catching that! fixed

File: test/e2e_node/swap_test.go (hunk 6 above)
Comment (iholder101) [reply test/e2e_node/swap_test.go]: Done!

File: test/e2e_node/swap_test.go (hunk 7 above)
Comment (iholder101) [reply test/e2e_node/swap_test.go:119]: >thank you for those. What swappiness value did you use in your configs? is it something we need to take into consideration.

Thanks for your review Sergey!

I've spent a lot of time writing these tests in a way that they will be as non-flaky and stable as possible, but at the same time will verify that every expectation from the NodeSwap feature is met. I've followed a few principles in order to achieve it:
* **The system is not configured in any special way**: The tests should work with default OS configurations to prove that NodeSwap is not dependent on any specific configuration like swappiness.
* **Tests end once the most basic functionality is verified**: For example, when ensuring that a container can exceed memory limits, the container (slowly but surely) stresses memory and once `memory.current > limits` and `memory.swap.current > 0` the test passes.

> So depending on how fast we write memory, test may start flaking.

Bear in mind that the stress is being made at a low pace to give the kernel enough time to reclaim. It is proven to be very stable until now, but we can always make it even slower if we'll see future flakiness.

I think that swappiness shouldn't matter too much here as well. Essentially it is a way to tell the kernel to prefer reclaiming file-backed memory vs anon memory, but in any case after enough stress the kernel would have to start swapping.

> The soft memory limit in cgroupv2 is something that I think may improve swap reliability.

I think you're referring to `memory.swap.high`.
This is indeed an interesting topic which I think about a lot. I'm not sure if it's in scope for this PR, but I would love discussing it offline if you have the time/interest.

**Comment by iholder101** [comment]
@SergeyKanzhelev @mrunalp
Can you PTAL?

**Review by mrunalp — APPROVED** [review]

**Comment by iholder101** [comment]
@dchen1107 @kannon92 can you PTAL?

File: test/e2e/framework/skipper/skipper.go [hunk 8]
```diff
@@ -67,6 +66,13 @@ func InitFeatureGates(defaults featuregate.FeatureGate, overrides map[string]boo
 	return nil
 }
 
+func IsFeatureGateEnabled(feature featuregate.Feature) bool {
```
Comment (pohly) [inline test/e2e/framework/skipper/skipper.go:69]: Can you added documentation? Feature gate testing is one of those areas that have been misunderstood by test authors in the past.

I'll re-add LGTM and approve when done.

/lgtm cancel

Comment (iholder101) [reply test/e2e/framework/skipper/skipper.go:72]: Thanks @pohly!
An important comment indeed. I'm also a victim of misunderstanding how it works :)

I've added the following documentation, let me know what you think:
> // IsFeatureGateEnabled can be used during e2e tests to figure out if a certain feature gate is enabled.
> // This function is dependent on InitFeatureGates under the hood. Therefore, the test must be called with a
> // --feature-gates parameter.
